### PR TITLE
[DO NOT MERGE]: Add generic combobox hooks

### PR DESCRIPTION
**Pull request by sopranopillow**
## Related Issue(s)

- Fixes #

### CODE REVIEW

File: packages/react-components/react-combobox/library/src/components/Listbox/useListbox.ts [hunk 1]
```diff
@@ -169,7 +169,7 @@ export const useListbox_unstable = (props: ListboxProps, ref: React.Ref<HTMLElem
     }
 
     // otherwise start at the first option
-    else {
+    else if (!disableAutoFocus) {
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/components/Listbox/useListbox.ts:172]: Wrap if above and current one on `!disableAutoFocus`

File: packages/react-components/react-combobox/library/src/components/Listbox/Listbox.types.ts [hunk 2]
```diff
@@ -16,7 +16,15 @@ export type ListboxSlots = {
...
  */
-export type ListboxProps = ComponentProps<ListboxSlots> & SelectionProps;
+export type ListboxProps = ComponentProps<ListboxSlots> &
+  SelectionProps & {
+    /**
+     * Disable autofocus on the first item on mount.
+     *
+     * @default false
+     */
+    disableAutoFocus?: boolean;
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/components/Listbox/Listbox.types.ts:26]: Add this to FUI

File: packages/react-components/react-combobox/library/src/contexts/usePromptInputContextValues.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import * as React from 'react';
+import { PromptInputState } from '../PromptInput';
+import { ComboboxBaseContextValues, ComboboxBaseState } from '../utils/ComboboxBase.types';
+
+export function usePromptInputContextValues(
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/contexts/usePromptInputContextValues.ts:5]:
```
useGenericInputContextValues = (state) => {
  useComboboxInputcontextValue({...state, appearance: 'outline' });
}
```

Comment (sopranopillow) [reply packages/react-components/react-combobox/library/src/contexts/usePromptInputContextValues.ts:5]: Repeat code and strip deprecated props, basically keep active descendant and necessary input stuff like setOpen

Comment (sopranopillow) [reply packages/react-components/react-combobox/library/src/contexts/usePromptInputContextValues.ts:5]: This should be in collab

File: packages/react-components/react-combobox/library/src/utils/useGenericComboboxBaseState.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  | 'inlinePopup'
+  | 'clearable'
+  | 'mountNode'
+> & {
+  children?: React.ReactNode;
+  editable?: boolean;
+  disabled?: boolean;
+  freeform?: boolean;
+  activeDescendantController: ActiveDescendantImperativeRef;
+};
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/utils/useGenericComboboxBaseState.ts:34]: just omit appearance, size, and deprecated stuff

File: packages/react-components/react-combobox/library/src/utils/useListboxSlot.ts [hunk 5]
```diff
@@ -19,7 +19,8 @@ type UseListboxSlotOptions = {
   triggerRef:
     | React.RefObject<HTMLInputElement>
     | React.RefObject<HTMLButtonElement>
-    | React.RefObject<HTMLButtonElement | HTMLInputElement>;
+    | React.RefObject<HTMLSpanElement>
+    | React.RefObject<HTMLButtonElement | HTMLInputElement | HTMLSpanElement>;
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/utils/useListboxSlot.ts:23]: Add HTMLDivElement as well

File: packages/react-components/react-combobox/library/src/utils/useTriggerSlot.ts [hunk 6]
```diff
@@ -139,63 +165,69 @@ function useTriggerKeydown(
...
+          setOpen(e, true);
+          break;
+        case 'Close':
+          // stop propagation for escape key to avoid dismissing any parent popups
+          e.stopPropagation();
+          setOpen(e, false);
+          break;
+        case 'CloseSelect':
+          !multiselect && !activeOption?.disabled && setOpen(e, false);
+        // fallthrough
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/utils/useTriggerSlot.ts:223]: do the same as useInputTriggerSlot that saves the onKeyDown.

File: packages/react-components/react-combobox/library/src/utils/useTriggerSlot.ts [hunk 7]
```diff
@@ -57,25 +66,42 @@ export function useTriggerSlot(
...
-    setHasFocus(false);
-  }, trigger.onBlur);
+  trigger.onBlur = mergeCallbacks(
+    (
+      event: React.FocusEvent<HTMLButtonElement> &
+        React.FocusEvent<HTMLInputElement> &
+        React.FocusEvent<HTMLSpanElement>,
+    ) => {
+      setOpen(event, false);
+      setHasFocus(false);
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/utils/useTriggerSlot.ts:82]: Merge inputtriggerSlot and triggerSlot into a custom triggerslot.

File: packages/react-components/react-combobox/library/src/index.ts [hunk 8]
```diff
@@ -59,13 +59,40 @@ export {
...
+  SelectionEvents,
+  SelectionProps,
+  SelectionState,
+  UseListboxSlotState,
+  UseTriggerSlotState,
+} from './utils/index';
+export {
+  getDropdownActionFromKey,
+  iconSizes,
+  useComboboxBaseState,
```
Comment (sopranopillow) [inline packages/react-components/react-combobox/library/src/index.ts:90]: Only export this

**Comment by layershifter** [comment]
Closing for housekeeping 🏡  Feel free to re-open

**Pull request closed without merging**